### DISCUSSION TITLE
Nullable list of structs in json document

### DISCUSSION THREAD

**Discussion by zlesnr** [cuelang/cue#821] [Q&A]
Howdy folks, I'm trying to create a "list" that is optional.  Running vet on a json document that contains null where the cue spec defines a list gives me a type mismatch error that I'm confused about how to resolve.  For example:
```
processes: [...#MatchedProcess]
```
In cue, while the json has
```
"processes": null
```

How can I have process either conform to the definition OR be null in order to get null to pass?

If I change the cue to look like so: 

```
processes:       [...#MatchedProcess] | null
```
Then the error I receive is `conflicting values null and [...string] (mismatched types null and list)`.

**Comment by mpvl** [comment]
This works for me:
```
processes: [...#MatchedProcess] | null
processes: null
#MatchedProcess: string
```
Can you describe the exact steps that lead to your error?

**Reply by zlesnr** [reply]
I'm unable to reproduce this this morning after getting it working yesterday.
```
❯ cat document_sample.json
{
  "discoveryManifest": {
    "processes": null
  }
}
❯ cat document.cue
discoveryManifest: #DiscoveryManifest

#DiscoveryManifest: {
        processes: [...#MatchedProcess] | null
}

#MatchedProcess: {
        command: string
}
```

Now `cue vet document_sample.json document.cue` executes without error.

I believe my confusion was that there were multiple errors in my document.  The simple nullable string has been updated to `entityGuid?: string`, where the `?` was missing before.

Is the `?` an alternate syntax for `|null`, or is the `|null` syntax required for structs, while the `?` is used on simple types?  I also see references to `|*null`, but I'm unclear on the distinction.

**Reply by mpvl** [reply]
`null` has no special meaning in CUE. It is there mostly for JSON compatibility. The usage of `null` in JSON is so polluted that it really wasn't possible to assign any meaning to it consistently. So in the end, it is just another value at your disposal, just like `true` or `false`, but without assigning any meaning.

`foo?: int` is a shorthand for `["foo"]: int`, but not for anything else.

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/821.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
